src/main/scala/nodes/learning/KernelGenerator.scala

```diff
@@ -0,0 +1,273 @@
+        rhsIdx += 1
+      } else if (lhsInds(lhsIdx) < rhsInds(rhsIdx)) {
+        lhsIdx += 1
+      } else {
+        rhsIdx += 1
+      }
+    }
+    s
+
+  }
+
+
+  def generateKernelTestBlock(
+    lhsMat: RDD[(Array[Long], Array[Double])],
+    rhsMat: RDD[(Array[Long], Array[Double])],
+    blockIdxs: Seq[Int]): RDD[DenseVector[Double]] = {
+
+    val blockDataArray = rhsMat.zipWithIndex.filter { case (_, idx) =>
+      blockIdxs.contains(idx)
+    }.map(x => x._1).collect()
+    val blockDataBC = rhsMat.context.broadcast(blockDataArray)
+    val kBlock = lhsMat.map { case (lhsInds, lhsValues) =>
+      new DenseVector(blockDataBC.value.map { case (rhsInds, rhsValues) =>
+        sparseDotProduct(lhsInds, lhsValues, rhsInds, rhsValues)
+      }.toArray)
+    }
+    kBlock.cache()
+    kBlock.count
+    blockDataBC.unpersist(true)
+    kBlock
+  }
+
+  def generateKernelTrainBlock(
+    mat: RDD[(Array[Long], Array[Double])],
+    blockIdxs: Seq[Int]): (RDD[DenseVector[Double]], DenseMatrix[Double]) = {
+
+    mat.foreach { x => assert(KernelUtils.isSorted(x._1)) } // limitation for now
+
+    val lhsMat = mat
+    val rhsMat = mat
+
+    val blockDataArray = rhsMat.zipWithIndex.filter { case (_, idx) =>
+      blockIdxs.contains(idx)
+    }.map(x => x._1).collect()
+    val blockDataBC = rhsMat.context.broadcast(blockDataArray)
+    val Kblock = lhsMat.map { case (lhsInds, lhsValues) =>
+      new DenseVector(blockDataBC.value.map { case (rhsInds, rhsValues) =>
+        sparseDotProduct(lhsInds, lhsValues, rhsInds, rhsValues)
+      }.toArray)
+    }
+
+    val Kblockblock = blockDataArray.par.map { case (lhsInds, lhsValues) =>
+      new DenseVector(blockDataArray.map { case (rhsInds, rhsValues) =>
+        sparseDotProduct(lhsInds, lhsValues, rhsInds, rhsValues)
+      }.toArray)
+    }.toArray
+
+    // val Kblockblock = blockDataArray.map { case (lhsInds, lhsValues) =>
+    //   new DenseVector(blockDataArray.map { case (rhsInds, rhsValues) =>
+    //     sparseDotProduct(lhsInds, lhsValues, rhsInds, rhsValues)
+    //   }.toArray)
+    // }
+
+    Kblock.cache()
+    Kblock.count
+    blockDataBC.unpersist(true)
+
+    (Kblock, KernelUtils.rowsToMatrix(Kblockblock))
+  }
+
+}
+
+class GaussianKernelGenerator(gamma: Double)
+  extends KernelGenerator[DenseVector[Double]]
+  with Serializable {
+
+  def generateKernelTestBlock(
+    testMat: RDD[DenseVector[Double]],
+    trainMat: RDD[DenseVector[Double]],
+    blockIdxs: Seq[Int]): RDD[DenseVector[Double]]= {
+
+    println("testMat.count=" + testMat.count +
+      ",trainMat.count=" + trainMat.count +
+      ", blockSize=" + blockIdxs.length)
+
+    // Dot product of rows of X with each other
+    val trainDotProd = trainMat.map {x =>
+      (x.t*x).toDouble
+    }
+    val testDotProd = testMat.map {x =>
+      (x.t*x).toDouble
+    }
+
+    // b x d block of training data
+    val blockDataArray = trainMat.zipWithIndex.filter{ case (vec, idx) =>
+      blockIdxs.contains(idx)
```

Comment:
this could get slow if someone chooses a big block size. You might want to make blockIdxs a "Set" if order is unimportant - or convert locally to a set for this kidn of check.